packages/bsky/src/subscription/backfill.ts

```diff
@@ -0,0 +1,149 @@
+import PQueue from 'p-queue'
+import { AtpAgent } from '@atproto/api'
+import { Subscription } from '@atproto/xrpc-server'
+import { OutputSchema as Message } from '../lexicon/types/com/atproto/sync/subscribeRepos'
+import { ids, lexicons } from '../lexicon/lexicons'
+import AppContext from '../context'
+import { subLogger } from '../logger'
+import { retryHttp } from '../util/retry'
+
+const METHOD = ids.ComAtprotoSyncSubscribeRepos
+
+export const backfillRepos = async (ctx: AppContext, concurrency: number) => {
+  if (!ctx.cfg.repoProvider) {
+    throw new Error('No repo provider for backfill')
+  }
+  // first, peek the stream to find the last event
+  const cursor = await peekStream(ctx)
+  if (cursor === null) {
+    subLogger.info('already caught up, skipping backfill')
+    console.log('SKIPPING backfill')
+    return
+  }
+
+  // then run backfill
+  await doBackfill(ctx, concurrency)
+
+  // finally update our subscription state to reflect the fact that we're caught up with the previously peeked cursor
+  const state = JSON.stringify({ cursor })
+  await ctx.db.db
+    .insertInto('subscription')
+    .values({
+      service: ctx.cfg.repoProvider,
+      method: METHOD,
+      state,
+    })
+    .onConflict((oc) =>
+      oc.columns(['service', 'method']).doUpdateSet({ state }),
+    )
+    .execute()
+}
+
+export const peekStream = async (ctx: AppContext): Promise<number | null> => {
+  const repoProvider = ctx.cfg.repoProvider
+  if (!repoProvider) {
+    throw new Error('No repo provider for backfill')
+  }
+
+  const sub = new Subscription({
+    service: repoProvider,
+    method: METHOD,
+    validate: (val) => {
+      return lexicons.assertValidXrpcMessage<Message>(METHOD, val)
+    },
+    getParams: async () => {
+      const lastSeenCursor = await ctx.db.db
+        .selectFrom('subscription')
+        .where('service', '=', repoProvider)
+        .where('method', '=', METHOD)
+        .selectAll()
+        .executeTakeFirst()
+      return lastSeenCursor ? JSON.parse(lastSeenCursor.state) : { cursor: 0 }
+    },
+  })
+  const evts = sub[Symbol.asyncIterator]()
+  const first = await evts.next()
+  // first message should be an OutdatedCursor info msg
+  if (first.done || first.value.name !== 'OutdatedCursor') {
+    return null
+  }
```

Comment:
Alternately we could ensure to never use params with the subscription here, and take the sequence number of the first event that comes through.